### ISSUE TITLE
Incorrect error reported for class that has decorator

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#592]
When using ipywidgets package.

```
import ipywidgets as widgets
out = widgets.Output(layout={'border': '1px solid black'})
```

**Comment by erictraut** [comment]
I investigated this further, and I don't see a good way to handle this case. It's a nasty bit of code in a decorator function that is intended to support two very different forms of the decorator. One form takes a parameter (`@register(name)`) and one doesn't (`@register`). The code dynamically determines which of these two cases to use, and it returns either a function or an object depending on the case. The type follows both paths but can't determine statically which one is going to be taken. The error that it reports is legitimate in light of this.

This is a case where a type stub is really needed to handle this correctly.

**Comment by erictraut** [comment]
For the record, here's the code in question:
```
def register(name=''):
    "For backwards compatibility, we support @register(name) syntax."
    def reg(widget):
        """A decorator registering a widget class in the widget registry."""
        w = widget.class_traits()
        Widget.widget_types.register(w['_model_module'].default_value,
                                    w['_model_module_version'].default_value,
                                    w['_model_name'].default_value,
                                    w['_view_module'].default_value,
                                    w['_view_module_version'].default_value,
                                    w['_view_name'].default_value,
                                    widget)
        return widget
    if isinstance(name, string_types):
        import warnings
        warnings.warn("Widget registration using a string name has been deprecated. Widget registration now uses a plain `@register` decorator.", DeprecationWarning)
        return reg
    else:
        return reg(name)
```

**Comment by erictraut** [comment]
@jakebailey Let me know if you have any ideas here. My inclination is to close this "as designed".

**Comment by jakebailey** [comment]
Why isn't this able to be determined statically? At the decoration site, we know that it's a function type, and if it's being passed in to be evaluated (the specialization), wouldn't we be able to pick the correct `isinstance` leg given we know that it's not a string? Or is the issue strictly with the `string_types` itself?

**Comment by erictraut** [comment]
Yes, `string_types` is a variable. If I replace it with `(str,)` (which is the value it contains), the static analyzer able to statically determine (through the call-site return type inference logic) the correct return type.

**Comment by jakebailey** [comment]
Have you thought about preserving values for very simple values like this? Similarly to what I believe you have to do already for `__all__`, for example.

**Comment by erictraut** [comment]
I'm not sure what you mean. The type analyzer doesn't track values, it tracks types. The closest that it comes to tracking values is when it interprets a literal value as a constrained type.

**Comment by jakebailey** [comment]
Here is something MPLS may currently do:

```python
x = 1234
y = "some string"
z = ["foo", "bar"]
```

When it walks this, it will end up with the state:

- `x` is a variable containing an instance of `int` with the value `1234`
- `y` is a variable containing an instance of `str` with the value `"some string"`
- `z` is an instance of `list` containing the two strings `"foo"` and `"bar"`"

When used later, the analysis may be able to make use of these values, or eventually throw them away.

This is how we perform more advanced operations on `__all__`, as we'll choose not to throw away this data and instead allow `+=`, `.append`, etc to modify it. Then, when we go looking for `__all__` later it's been populated with values. If something more complicated happens, then we'll throw it away and we'll just say it's a `list` with unknown contents.

When I say preserving values, I mean this sort of thing, where you can definitively say that some variable has some value and use it later.

I can see how this wouldn't be as favorable in pyright where you really only want to be talking in terms of types and not values (and, potentially incorrect if say some outside code were to push onto that list and change it, or something).

`Literal` is the closest thing I can think of to this, but I don't think you can have a literal tuple/list value (as nice as that'd be, actually, given things like `isinstance` or `super` which make use of lists of types for checks, MRO ordering, others).

**Comment by jakebailey** [comment]
Sidenote: jedi's analysis can even fully statically evaluate expressions like `getattr(x, "".join(reversed(["r", "a", "b", "o", "o", "f"])))` as `x.foobar`. Perhaps overkill, but an example of a similar thing.

**Comment by erictraut** [comment]
Yeah, I don't think the type checker should be in the business of tracking values. That's the job of an interpreter. :)

This is a good example of a problem that can and should be solved via a type stub.

**Comment by jakebailey** [comment]
Perhaps, but Python isn't TypeScript and types are values that can happily be passed around as they really do exist (and often are in libraries that want to do fancy things). I think there's going to be some bleedthrough between type checking and value tracing, and this decorator is an example where it was convenient for the code author to create a variable with string types to use in multiple places, as `isinstance` does not care that you wrote them inline. It just wanted something to range over.

The good news for this function is that a later version replaced the variable with `isinstance(name, str)`, since Python 3 only has `str` and not `unicode`. Then, in jupyter-widgets/ipywidgets#2695 the extra form got removed entirely. Maybe people should just update their environments to get new code... 😃 (Of course, scanning the file, I see more places where this has been done, so it's not all good news.)
